File: test_shared/test_runner.go
package main

import (
	"fmt"
	"strings"
)

func RunTest[T any, S any](
	problemName string,
	testCases []T, 
	test func(testCase T, solution S) (bool, error),
	solution S,
	) {
	passed := 0
	failed := 0

	fmt.Println(fmt.Sprintf("Running Tests for: %s", problemName))
	fmt.Println(strings.Repeat("=", 70))

	for i, testCase := range testCases {
		result, err := test(testCase, solution)

		if err != nil {
			failed++
			fmt.Println(fmt.Sprintf("Test %d failed", i+1))
			fmt.Println(fmt.Sprintf("   Error: %s", err.Error()))
		}

		if result {
			passed++
			fmt.Println(fmt.Sprintf("Test %d passed", i+1))
		}
	}

	fmt.Println(strings.Repeat("=", 70))
	fmt.Println(fmt.Sprintf("Results: %d passed, %d failed, %d total", passed, failed, len(testCases)))

	if failed == 0 {
		fmt.Println("All tests passed!")
	}
}

File: test_shared/test_runner.rs
use std::io;

pub fn run_test<T,S: Copy>(problem_name: &str, test_cases: Vec<T>, test: fn(&T, S) -> Result<bool, io::Error>, solution: S) {
    let mut passed = 0;
    let mut failed = 0;

    println!("Running tests for: {}", problem_name);
    println!("{}", "=".repeat(70));

    for (i, test_case) in test_cases.iter().enumerate() {
        let result = test(test_case, solution);
        match result {
            Ok(true) => {
                passed += 1;
                println!("Test {} passed", i);
            }
            Ok(false) => {
                failed += 1;
                println!("Test {} failed", i);
            }   
            Err(err) => {
                failed += 1;
                println!("Test {} failed", i);
                println!("   Error: {}", err);
            }
        }
    }

    println!("{}", "=".repeat(70));
    println!("Results: {} passed, {} failed, {} total", passed, failed, test_cases.len());

    if failed == 0 {
        println!("All tests passed!");
    }

}